core/src/main/java/io/grpc/internal/ProxyDetectorImpl.java

```diff
@@ -0,0 +1,178 @@
+      new Supplier<ProxySelector>() {
+        @Override
+        public ProxySelector get() {
+          // TODO(spencerfang): consider using java.security.AccessController here
+          return ProxySelector.getDefault();
+        }
+      };
+
+  @Deprecated
+  private static final String GRPC_PROXY_ENV_VAR = "GRPC_PROXY_EXP";
+  // Do not hard code a ProxySelector because the global default ProxySelector can change
+  private final Supplier<ProxySelector> proxySelector;
+  private final AuthenticationProvider authenticationProvider;
+  private final ProxyParameters override;
+
+  // We want an HTTPS proxy, which operates on the entire data stream (See IETF rfc2817).
+  static final String PROXY_SCHEME = "https";
+  static final String URI_FORMAT = PROXY_SCHEME + "://%s";
+
+  /**
+   * A proxy selector that uses the global {@link ProxySelector#getDefault()} and
+   * {@link ProxyDetectorImpl.AuthenticationProvider} to detect proxy parameters.
+   */
+  public ProxyDetectorImpl() {
+    this(DEFAULT_PROXY_SELECTOR, DEFAULT_AUTHENTICATOR, System.getenv(GRPC_PROXY_ENV_VAR));
+  }
+
+  @VisibleForTesting
+  ProxyDetectorImpl(
+      Supplier<ProxySelector> proxySelector,
+      AuthenticationProvider authenticationProvider,
+      @Nullable String proxyEnvString) {
+    this.proxySelector = proxySelector;
+    this.authenticationProvider = authenticationProvider;
+    if (proxyEnvString != null) {
+      override = new ProxyParameters(overrideProxy(proxyEnvString), null, null);
+    } else {
+      override = null;
+    }
+  }
+
+  @Nullable
+  @Override
+  public ProxyParameters proxyFor(SocketAddress targetServerAddress) {
+    if (override != null) {
+      return override;
+    }
+    if (!(targetServerAddress instanceof InetSocketAddress)) {
+      return null;
+    }
+    return detectProxy((InetSocketAddress) targetServerAddress);
+  }
+
+  private ProxyParameters detectProxy(InetSocketAddress targetAddr) {
+    String hostPort =
+        GrpcUtil.authorityFromHostAndPort(targetAddr.getHostName(), targetAddr.getPort());
+    URI uri;
+    try {
+      uri = new URI(String.format(URI_FORMAT, hostPort));
+    } catch (final URISyntaxException e) {
+      log.log(
+          Level.SEVERE,
+          "Failed to construct URI for proxy lookup, proceeding without proxy",
+          e);
+      return null;
+    }
+
+    List<Proxy> proxies = proxySelector.get().select(uri);
+    if (proxies.size() > 1) {
+      log.warning("More than 1 proxy detected, gRPC will select the first one");
+    }
+    Proxy proxy = proxies.get(0);
+
+    if (proxy.type() == Proxy.Type.DIRECT) {
+      return null;
+    }
+    InetSocketAddress proxyAddr = (InetSocketAddress) proxy.address();
+    PasswordAuthentication auth = authenticationProvider.requestPasswordAuthentication(
+        proxyAddr.getHostName(),
+        proxyAddr.getAddress(),
+        proxyAddr.getPort(),
+        PROXY_SCHEME,
+        "proxy authentication required by gRPC",
```

Comment:
This is supposed to be the realm returned by the server. We won't have it since we are avoiding the full authentication handshake. We maybe should use `""`, but at the very least it deserves a comment.